include/cocaine/rpc/asio/header.hpp

```diff
@@ -0,0 +1,632 @@
+#include <cstring>
+
+namespace cocaine { namespace io  {
+namespace header {
+
+struct data_t {
+    const char* blob;
+    size_t size;
+    bool operator==(const data_t& other) const {
+        return size == other.size && (std::memcmp(blob, other.blob, size) == 0);
+    }
+
+    template<class To>
+    To
+    convert() {
+        static_assert(std::is_pod<typename std::remove_reference<To>::type>::value &&
+                      !std::is_pointer<typename std::remove_reference<To>::type>::value &&
+                      !std::is_array<typename std::remove_reference<To>::type>::value,
+                      "only POD non pointer, non array data type is allowed to convert header data"
+                      );
+        if(size != sizeof(typename std::remove_reference<To>::type)) {
+            throw cocaine::error_t("invalid header data size");
+        }
+        return *(reinterpret_cast<const To*>(blob));
+    }
+};
+
+template<size_t N>
+inline
+constexpr
+data_t
+create_data(char const (&source)[N]) {
+    return data_t{source, N-1};
+}
+
+template<class From>
+inline
+constexpr
+header::data_t
+create_data(From&& source) {
+    static_assert(std::is_lvalue_reference<From>::value &&
+                  std::is_pod<typename std::remove_reference<From>::type>::value &&
+                  !std::is_same<const char*, typename std::remove_reference<From>::type>::value,
+                  "only lreference to POD is allowed to create header data"
+                  );
+    return data_t{
+        reinterpret_cast<const char*>(&source),
+        sizeof(typename std::remove_reference<From>::type)
+    };
+}
+
+inline
+constexpr
+data_t
+create_data(const char* source, size_t size) {
+    return data_t{source, size};
+}
+} // namespace header
+
+struct headers;
+// Header class.
+// Represents non-owning header referring to some external memory.
+// Can be constructed only via header table
+class header_t {
+public:
+    header_t(const header_t&) = default;
+    header_t& operator=(const header_t&) = default;
+    header_t(): name(), value() {}
+
+    // Create non-owning header on user-provided data
+    template<class Header>
+    static
+    header_t
+    create(header::data_t _value) {
+        return header_t(Header::name, _value);
+    }
+
+    inline
+    header::data_t
+    get_name() const;
+
+    inline
+    header::data_t
+    get_value() const;
+
+    bool
+    operator==(const header_t& other) const {
+        return name == other.name && value == other.value;
+    }
+
+    bool
+    name_equal(const header_t& other) const {
+        return name == other.name;
+    }
+
+    // Returns size of the header entry in http2 header table
+    inline
+    size_t
+    http2_size() const;
+
+    friend struct header_traits;
+    friend struct io::headers;
+    friend struct header_static_table_t;
+
+private:
+    friend class header_table_t;
+
+    inline
+    header_t(const header::data_t& _name, const header::data_t& _value) noexcept;
```

Comment:
Do not mangle argument names (like `_name`) in interface declaration please, even if it is private.